sound/soc/sof/sof-audio.c

```diff
@@ -213,65 +285,245 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
 int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
 {
 	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	int num_pipe_widgets = 0;
+	int i, ret, num_widgets;
 
 	/* nothing to set up */
 	if (!list)
 		return 0;
 
-	return sof_widget_list_dai_config_setup(sdev, list);
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, num_widgets, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			goto widget_free;
+	}
+
+	/*
+	 * error in setting pipeline connections will result in route status being reset for
+	 * routes that were successfully set up when the widgets are freed.
+	 */
+	ret = sof_setup_pipeline_connections(sdev, list, dir);
+	if (ret < 0)
+		goto widget_free;
+
+	/* setup pipeline widgets and complete pipelines */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget;
+
+		if (!swidget)
+			continue;
+
+		pipe_widget = swidget->pipe_widget;
+		if (!pipe_widget) {
+			dev_err(sdev->dev, "error: no pipeline widget found for %s\n",
+				swidget->widget->name);
+			ret = -EINVAL;
+			goto pipe_free;
+		}
+
+		ret = sof_widget_setup(sdev, pipe_widget);
+		if (ret < 0)
+			goto pipe_free;
+
+		num_pipe_widgets++;
+
+		if (pipe_widget->complete)
+			continue;
+
+		pipe_widget->complete = snd_sof_complete_pipeline(sdev->dev, pipe_widget);
+		if (pipe_widget->complete < 0) {
+			ret = pipe_widget->complete;
+			goto pipe_free;
+		}
+	}
+
+	return 0;
+
+pipe_free:
+	/* free the pipe widgets that were set up successfully */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (!num_pipe_widgets--)
+			break;
+
+		sof_widget_free(sdev, swidget->pipe_widget);
+	}
+
+widget_free:
+	/* free all widgets that have been set up successfully */
+	for_each_dapm_widgets(list, i, widget) {
+		if (!num_widgets--)
+			break;
+
+		if (is_sof_widget(widget))
+			sof_widget_free(sdev, widget->dobj.private);
+	}
+
+	return ret;
 }
 
-/* placeholder for now. Will be updated in the following patches */
 int sof_widget_list_free(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
 {
 	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	int i, ret;
 
 	/* nothing to free */
 	if (!list)
 		return 0;
 
-	return sof_widget_list_dai_config_setup(sdev, list);
+	/*
+	 * Free widgets in the list. This can fail but continue freeing other widgets to keep
+	 * use_counts balanced.
+	 */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		/* free widget and its pipe_widget */
+		ret = sof_widget_free(sdev, swidget);
+		if (ret < 0)
+			return ret;
```

Comment:
if one widget is failed to be free, do we need to free other widgets ? And also the following pipeline

Reply:
yes, we should free as many a spossible and free the list before returning the error. fixed now